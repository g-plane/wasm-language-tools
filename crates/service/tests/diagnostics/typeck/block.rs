use super::*;
use insta::assert_json_snapshot;
use lsp_types::Uri;
use wat_service::LanguageService;

#[test]
fn block_type_in_stack() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (func (param i32 i32) (result i32)
    (block (result i32 i32)
      (local.get 0)
      (local.get 1))
    i32.add))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert!(pick_diagnostics(response).is_empty());
}

#[test]
fn block_type_folded() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (func (param i32 i32) (result i32)
    (i32.add
      (block (result i32 i32)
        (local.get 0)
        (local.get 1)))))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert!(pick_diagnostics(response).is_empty());
}

#[test]
fn folded() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (func (param i32 f32) (result i32)
    (block (result i32)
      (i32.add
        (local.get 0)
        (local.get 1)))))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn sequence() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (func (param i32 f32) (result i32)
    block (result i32)
      local.get 0
      local.get 1
      i32.add
    end))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn new_stack_for_new_block() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (func (result i32)
    i32.const 0
    i32.const 1
    block (result i32)
      i32.add
    end))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn params_boundary_missing() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (func
    i32.const 0
    block (param i32 i32)
      i32.add
      drop
    end))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn params_boundary_mismatched() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (func
    i32.const 0
    f32.const 1
    block (param i32 i32)
      i32.add
      drop
    end))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn params_mismatched() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (func
    i32.const 0
    f32.const 1
    block (param i32 f32)
      i32.add
      drop
    end))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn params_correct() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (func
    i32.const 0
    i32.const 1
    block (param i32 i32)
      i32.add
      drop
    end))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert!(pick_diagnostics(response).is_empty());
}

#[test]
fn results_folded() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (func (result i32 i32)
    (block (result i32 i32)
      (i32.const 0)))
  (func (result i32 i32)
    (block (result i32 i32)
      (i32.const 0)
      (f32.const 0)))
  (func (result i32 i32)
    (block (result i32 i32)
      (i32.const 0)
      (i32.const 0)
      (i32.const 0)))
  (func (result i32 i32)
    (block (result i32 i32)
      (i32.const 0)
      (i32.const 0)
      (f32.const 0))))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn results_sequence() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (func (result i32 i32)
    block (result i32 i32)
      i32.const 0
    end)
  (func (result i32 i32)
    block (result i32 i32)
      i32.const 0
      f32.const 0
    end)
  (func (result i32 i32)
    block (result i32 i32)
      i32.const 0
      i32.const 0
      i32.const 0
    end)
  (func (result i32 i32)
    block (result i32 i32)
      i32.const 0
      i32.const 0
      f32.const 0
    end))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn results_correct() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (type $t (func (result i32 i32)))
  (func (result i32 i32)
    (block (result i32 i32)
      unreachable))
  (func (result i32 i32)
    (block (result i32 i32)
      (i32.const 0)
      (f32.const 0)
      (unreachable)))
  (func (result i32 i32)
    (block (result i32 i32)
      (f32.const 0)
      (i32.const 0)
      (unreachable)))
  (func (result i32 i32)
    (block (result i32 i32)
      (i32.const 0)
      (i32.const 0)))
  (func (result i32 i32)
    (block (type $t)
      (i32.const 0)
      (i32.const 0)))
  (func (result i32 i32)
    block (result i32 i32)
      unreachable
    end)
  (func (result i32 i32)
    block (result i32 i32)
      i32.const 0
      f32.const 0
      unreachable
    end)
  (func (result i32 i32)
    block (result i32 i32)
      f32.const 0
      i32.const 0
      unreachable
    end)
  (func (result i32 i32)
    block (result i32 i32)
      i32.const 0
      i32.const 0
    end))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert!(pick_diagnostics(response).is_empty());
}
