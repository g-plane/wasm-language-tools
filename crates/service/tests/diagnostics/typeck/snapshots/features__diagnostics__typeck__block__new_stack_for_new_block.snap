---
source: crates/service/tests/diagnostics/typeck/block.rs
expression: response
---
{
  "kind": "full",
  "items": [
    {
      "range": {
        "start": {
          "line": 6,
          "character": 6
        },
        "end": {
          "line": 6,
          "character": 13
        }
      },
      "severity": 1,
      "code": "type-check",
      "source": "wat",
      "message": "expected types [i32, i32], found []"
    },
    {
      "range": {
        "start": {
          "line": 7,
          "character": 7
        },
        "end": {
          "line": 7,
          "character": 8
        }
      },
      "severity": 1,
      "code": "type-check",
      "source": "wat",
      "message": "expected types [i32], found [i32, i32, i32] at the end"
    }
  ]
}
